### ISSUE TITLE
Global beforeEach/afterEach

### ISSUE THREAD

**Issue by brapifra** [dubzzz/fast-check#876]
## 🚀 Feature Request

Similarly to test runners like `jest`, it would be great if there was a way of setting global beforeEach/afterEach functions that get executed before/after a property is run.

## Motivation

When using libraries like react-testing-library, it is recommended to call its `cleanup` method between tests to prevent issues like memory leaks. While this is done automatically for popular test runners, I need to manually do it for fast-check tests.
https://testing-library.com/docs/react-testing-library/api#cleanup

## Example
Currently, I have to do this:
```typescript
fc.property(...).beforeEach(cleanup);
```
Ideally, I would want to do something like this:

```typescript
// In jest.setup.ts file
fc.configureGlobal({
    beforeEach: cleanup,
});
// Then I could skip the `beforeEach` from properties
fc.property(...);
```

**Comment by dubzzz** [comment]
Indeed it would be a pretty cool and interesting feature to bring into fast-check. Actually even if for the moment `configureGlobal` is only used to store configuration settings that would have been passed towards `assert` or `check` I have already plans to extend it to more options like the one you suggested here (see https://github.com/dubzzz/fast-check/issues/452).

To go back on the suggested feature: it is clearly something that would be great to have built-in.

A suggestion would be to add it into the typings of [`GlobalParameters`](https://github.com/dubzzz/fast-check/blob/101eb97228b5e08bb0c0d5855e0fccbf0a6d4ccf/src/check/runner/configuration/GlobalParameters.ts#L6) with something like:

```ts
type GlobalParameters = Pick<Parameters...> & { otherOptionsNotUsedByAssert };
```

As for [`reporter`](https://github.com/dubzzz/fast-check/blob/101eb97228b5e08bb0c0d5855e0fccbf0a6d4ccf/src/check/runner/configuration/Parameters.ts#L116-L136) setting we would have to deal with two possible versions: one synchronous and the other one asynchronous. So it might require 4 parameters (2 for synchronous, 2 for asynchronous). [As for `reporter`](https://github.com/dubzzz/fast-check/blob/101eb97228b5e08bb0c0d5855e0fccbf0a6d4ccf/src/check/runner/configuration/Parameters.ts#L133-L134) I would say that if one of the asynchronous method is set asynchronous properties will be forbidden.

...

Anyway, let's go for such feature! Would you be interested into giving it a try?

I can clearly send you pointers for that if needed.

**Comment by brapifra** [comment]
Great! I'm happy to give it a try.
Why would it require 4 parameters though? Couldn't we just have 2 that handle synchronous and asynchronous callbacks?
```typescript
type HookFunction = () => (Promise<void> | void);
beforeEach(callback: HookFunction) {
    ...
}

run() {
    ...
    await beforeEach(); // If it's a synchronous fn, it will return immediately
    ...
}
```

**Comment by dubzzz** [comment]
Unfortunately, synchronous code canot wait for asynchronous code to end.

The code below is invalid and will not execute as await cannot be used outside of async functions:
```js
run() {
  await beforeEach();
  classicRun();
  await afterEach();
}
```

Let's suppose we just remove the await in case of sync version of run but let users define beforeEach as they want. It's even worst as the beforeEach might trigger an asynchronous operation and the run will be executed before beforeEach ends. This is the reason why synchronous version of Property explicitly forbids to call beforeEach with async functions (either using async or returning a Promise).

**Comment by dubzzz** [comment]
In my opinion, the global setting should be read by the constructor of Property and AsyncProperty as soon as they get called. So that the behavior is identical to how global settings are used by assert.

**Comment by dubzzz** [comment]
> Great! I'm happy to give it a try.

Perfect 👌 Fill free to open a PR whenever ready or if you have any questions regarding a blocking point

**Comment by brapifra** [comment]
Oh okay, there are two versions of `run`, one async and other sync. Makes sense to have 4 options then.
I'll start looking into it as soon as I can! Thanks

**Comment by dubzzz** [comment]
Released with version 2.3.0 of fast-check.

Feel free to re-open :)
